test/e2e-auto-tls.sh

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env bash
+
+# Copyright 2020 The Knative Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+function setup_auto_tls_env_variables() {
+  # DNS zone for the testing domain.
+  export DNS_ZONE="knative-e2e"
+  # Google Cloud project that hosts the DNS server for the testing domain `kn-e2e.dev`
+  export CLOUD_DNS_PROJECT="knative-e2e-dns"
+  # The service account credential file used to access the DNS server.
+  export CLOUD_DNS_SERVICE_ACCOUNT_KEY_FILE="/etc/test-account/service-account.json"
+
+  export CUSTOM_DOMAIN_SUFFIX="$(($RANDOM % 10000)).${E2E_PROJECT_ID}.kn-e2e.dev"
+
+  local INGRESS_NAMESPACE=${GATEWAY_NAMESPACE}
+  if [[ -z "${GATEWAY_NAMESPACE}" ]]; then
+    INGRESS_NAMESPACE="istio-system"
+  fi
+  local INGRESS_SERVICE=${GATEWAY_OVERRIDE}
+  if [[ -z "${GATEWAY_OVERRIDE}" ]]; then
+    INGRESS_SERVICE="istio-ingressgateway"
+  fi
+  local IP=$(kubectl get svc -n ${INGRESS_NAMESPACE} ${INGRESS_SERVICE} -o jsonpath="{.status.loadBalancer.ingress[0].ip}")
+  export INGRESS_IP=${IP}
+}
+
+function setup_custom_domain() {
+  kubectl patch cm config-domain -n knative-serving -p "{\"data\":{\"${CUSTOM_DOMAIN_SUFFIX}\":\"\"}}"
+}
+
+function cleanup_custom_domain() {
+  kubectl apply -f ./config/config-domain.yaml
+}
+
+function turn_on_auto_tls() {
+  kubectl apply -f ./test/config/autotls/config-network.yaml
+}
+
+function turn_off_auto_tls() {
+  kubectl apply -f ./config/config-network.yaml
+}
+
+function setup_auto_tls_common() {
+  setup_auto_tls_env_variables
+
+  setup_custom_domain
+
+  # Disable namespace certificate.
+  unset NamespaceWithCert
+  go run ./test/e2e/autotls/config/disablenscert
```

Comment:
TBH, I'd rather not install it until we need it.  What if it leaves around wildcard certs in namespaces?

Reply:
OK. I updated the code to only install nscert controller when testing this scenario.